lib/src/main/scala/spinal/lib/blackbox/lattice/ecp5/EHXPLLL.scala

```diff
@@ -0,0 +1,224 @@
+package spinal.lib.blackbox.lattice.ecp5
+
+import spinal.core._
+
+object EHXPLLLConfig {
+
+  trait FeedbackType {
+    override def toString: String = ???
+  }
+  // Uses clock but expects this to be fed back in through CLKFB at the selected frequency
+  object CLKOP extends FeedbackType {
+    override def toString: String = "CLKOP"
+  }
+  object CLKOS extends FeedbackType {
+    override def toString: String = "CLKOS"
+  }
+  object CLKOS2 extends FeedbackType {
+    override def toString: String = "CLKOS2"
+  }
+  object CLKOS3 extends FeedbackType {
+    override def toString: String = "CLKOS3"
+  }
+  // Uses internal connections from the selected clock
+  object INT_CLKOP extends FeedbackType {
+    override def toString: String = "CLKOP"
+  }
+  object INT_CLKOS extends FeedbackType {
+    override def toString: String = "CLKOS"
+  }
+  object INT_CLKOS2 extends FeedbackType {
+    override def toString: String = "CLKOS2"
+  }
+  object INT_CLKOS3 extends FeedbackType {
+    override def toString: String = "CLKOS3"
+  }
+  object USER extends FeedbackType {
+    override def toString: String = "USERCLOCK"
+  }
+
+  private def VCO_MAX: HertzNumber = 800 MHz
+  private def VCO_MIN: HertzNumber = 400 MHz
+
+  /**
+    * Calculates the dividers required for the CLKOP path
+    * @param inFreq CLKI frequency
+    * @param outFreq CLKOP desired frequency
+    * @param tolerance Tolerance for error in CLKOP
+    * @return EHXPLLL configuration
+    */
+  def singleOutput(inFreq: HertzNumber, outFreq: HertzNumber, tolerance: BigDecimal = 0.02): EHXPLLLConfig = {
+    def calculateError(clki: BigDecimal, clko: BigDecimal, factor: BigDecimal): BigDecimal = {
+      (clko-(clki*factor))/clko
+    }
+
+    var mDiv: BigDecimal = 1
+    var fbDiv: BigDecimal = 1
+    var err: BigDecimal = calculateError(inFreq.toBigDecimal, outFreq.toBigDecimal, fbDiv/mDiv)
+    while(mDiv < 128 && fbDiv < 128 && err.abs > tolerance) {
```

Comment:
Would it make sense to add limit checks in this loop? Otherwise it could find a "best solution" that actually has an invalid internal frequency?

Reply:
Yes, it should attempt to keep the VCO within tolerance and valid frequency range.  The config will do the bounds check and it would be annoying if this function returns a VCO that's out of range.